src/System.Net.NameResolution/src/System/Net/DNS.cs

```diff
@@ -337,19 +314,26 @@ private static IAsyncResult HostResolutionBeginHelper(string hostName, bool just
             }
             else
             {
-                asyncResult = new ResolveAsyncResult(hostName, null, includeIPv6, state, requestCallback);
+                asyncResult = new DnsResolveAsyncResult(hostName, null, includeIPv6, state, requestCallback);
             }
 
             // Set up the context, possibly flow.
             asyncResult.StartPostingAsyncOp(false);
 
-            // Start the resolve.
-            Task.Factory.StartNew(
-                s => ResolveCallback(s),
-                asyncResult,
-                CancellationToken.None,
-                TaskCreationOptions.DenyChildAttach,
-                TaskScheduler.Default);
+            if (NameResolutionPal.SupportsGetAddrInfoAsync && includeIPv6 && SocketProtocolSupportPal.OSSupportsIPv6 && address == null)
```

Comment:
Should `NameResolutionPal.SupportsGetAddrInfoAsync` factor in `SocketProtocolSupportPal.OSSupportsIPv6` if that's part of what's required to use GetAddrInfoAsync?

Reply:
More generally, why is OSSupportsIPv6 necessary for SupportsGetAddrInfoAsync?

Reply:
GetAddrInfoEx doesn't require the OS to support IPv6. But the previous implementation was using GetHostByName instead of GetAddrInfo is OSSupportsIPv6 is false.

I'm not sure why that choice was made originally, but I thought it safe to change this behavior.

Reply:
Just saw the explanation in InternalGetHostByName

```
// Note        : Whilst getaddrinfo is available on WinXP+, we only
//               use it if IPv6 is enabled (platform is part of that
//               decision). This is done to minimize the number of
//               possible tests that are needed.
```

Reply:
Yeah, that comment is ancient.  There's no reason for us to not always use getaddrinfo today, and it would simplify the code and the tests.

My concern is mainly that we not make things any worse than they currently are.  So if you want to do the IPv6 checks, that's fine, but please add comments wherever you are doing so that refer back to the comment above, so it's clear why these checks are happening.

Reply:
I created issue #26856 to track this.